STREAMLINE-727 bootstrap-stroage script should allow create only and …

HeartSaVioR opened the pull request:
…keep deletion as an option

* bootstrap-storage.sh receives one optional argument
  * default value is "create"
  * "create": create tables
  * "drop": drop tables
  * "drop-create": drop then create tables
  * "check-connection": only check connection
* create new class StreamlineTableInitializer to keep SQLScriptRunner usable for general case
  * new class will handle only the case what bootstrap-storage provides
  * refactor SQLScriptRunner to let new class uses features

Regarding concurrent execution, since we're using `IF NOT EXISTS` for creating table and `IF EXISTS` for dropping table so it should be safe.

satishd commented on storage/tool/src/main/java/com/hortonworks/streamline/storage/tool/SQLScriptRunner.java:
```diff
@@ -39,44 +42,99 @@
     }
-    
+
+    public void downloadMySQLJarIfNeeded(String mysqlJarUrl) throws Exception {
```
I guess this class should not be specific to MySql. It is more specific to any JDBC supported storage. PR #514 already supports Postgres. It is better to clean this up.

HeartSaVioR replied:
Makes sense. I'll move out this method to new utility class, and let callers call that instead.

HeartSaVioR replied:
Btw, like below it shouldn't hurt to PostgreSQL storage setup.

harshach replied:
@satishd My understanding is postgres library is ASLv2 complaint where as mariadb or mysql driver is not. Hence we only added for mysql one. For postgres we should ship the library in classpath and shouldn't require a download.

satishd commented on bootstrap/bootstrap-storage.sh:
```diff
@@ -40,34 +41,60 @@ else
+}
+
+function createStreamlineTables {
+    ${JAVA} -Dstreamline.bootstrap.dir=$BOOTSTRAP_DIR  -cp ${CLASSPATH} ${STREAMLINE_TABLE_INITIALIZER_MAIN_CLASS} -m ${MYSQL_JAR_URL_PATH} -c ${CONFIG_FILE_PATH} --create-script ${SCRIPT_DIR}/create_tables.sql --create
```
This script should not be specific to MySql. We are going to support Postgres with #514 
You may want to make appropriate changes.

HeartSaVioR replied:
It is not that easy to make the change from here. The script doesn't know which storage type the configuration sets up. Previously the script knows about storage type (via executing another Java class), but we abstracted it to the Java class to reduce two operations into one.

It shouldn't hurt to PostgreSQL storage setup, too.

harshach replied:
@HeartSaVioR given that we have util class to read the streamline configs. We should use that figure out the db.type from config.

harshach replied:
Also we are expecting sql files to be under bootstrap/sql. Instead of passing all these options I would recommend we follow a syntax.
i.e store all the sql files in this manner
bootstrap/sql/db_type/create_tables.sql , bootstrap/sql/db_type/drop_tables.sql.
we can change the db_type mysql or postegres, phoenix etc..
Instead of having to pass drop_tables.sql or create_tables.sql . You only pass the parent directory under which sql scripts are stored. In this case, it would be "bootstrap/sql".
Your Java Class then can read the streamline  config to determine the right db.type and use that to invoke right db_type sql script and also it can take an argument such as "--create" or "--drop" like you are doing already but no need to pass the specific script to do so.

HeartSaVioR replied:
@harshach Yes new class is used only for bootstrapping so we could put the sql file placement information to the class. Great point. I'll apply it.

HeartSaVioR replied:
@harshach Addressed.

harshach replied:
@satishd @HeartSaVioR are we in agreement with the above approach.

harshach replied:
@HeartSaVioR thanks

satishd commented on bootstrap/bootstrap-storage.sh:
```diff
@@ -40,34 +41,60 @@ else
-    echo "Script files option: $FILE_OPT"
-    exec ${JAVA} -Dstreamline.bootstrap.dir=$BOOTSTRAP_DIR  -cp ${CLASSPATH} ${SCRIPT_RUNNER_MAIN_CLASS} -m ${MYSQL_JAR_URL_PATH} -c ${CONFIG_FILE_PATH} ${FILE_OPT}
+function checkStorageConnection {
+    ${JAVA} -Dstreamline.bootstrap.dir=$BOOTSTRAP_DIR  -cp ${CLASSPATH} ${STREAMLINE_TABLE_INITIALIZER_MAIN_CLASS} -m ${MYSQL_JAR_URL_PATH} -c ${CONFIG_FILE_PATH} --check-connection
```
Same as earlier comment.

HeartSaVioR replied:
Same here.

HeartSaVioR commented:
@satishd 
I might be misunderstanding your comment. Where do you think mysql specific logic should be placed?
Please let me know that you're thinking about moving out `mysql lib download` to the bootstrap script.

satishd commented:
@HeartSaVioR Respective storage type scripts like mysql and postgres can have their own scripts (`bootstrap-storage-<mysql/postgres>.sh`) which can share common functionality from other script like `bootstrap-storage-common.sh`.

harshach commented:
@satishd @HeartSaVioR we are waiting on this to unblock ambari changes and QE. I added my comments to make this generic. Let me know if you've any questions regarding this.

HeartSaVioR commented:
@satishd @harshach 
I feel we need to arrange this. Let's look at the history of change.

1. For the first time, this was implemented to two classes: config reader / sql runner. (IOT-441, #265)
Config reader reads config and figures out connection information, and sql runner runs the scripts for given dbtype.

2. After that we changed config reader to an utility class (hence not called by script), and utilize config reader class into sql runner. (STREAMLINE-612, #419) 

3. We needed downloading MySQL driver because of license issue, and added the feature: download MySQL driver to sql runner. (STREAMLINE-643, #449)
Before this change it was generic.

If we want to make the thing generic, let's just exclude mysql driver download to the outside of sql runner, and maybe, even outside of bootstrap-storage. Sql runner was generic before handling mysql driver download.
(Users might need to execute separate script for MySQL for the first time. If we think it hurts UX and don't want to do that, bootstrap-storage or sql runner should have it as exceptional case and can't be generic.)

What do you think?

harshach commented:
@HeartSaVioR mysql driver will be downloaded only when we configure the db.type to be "mysql". why we need to change that?

HeartSaVioR commented:
@harshach 
What @satishd commented about moving out specific implementation is all related to mysql download. (Please look through where @satishd left comments.)
So my understanding of review comments are allowing non-generic vs generic and move out mysql driver download. It doesn't represent that I want to move out the logic, just thinking about the way to make it db-type generic. Either (keep as it is or change it) is fine for me.

@satishd Please follow up the thread and let us know your thought.

satishd commented:
@HeartSaVioR @harshach I am fine with the change as if it is one off the case.

satishd commented:
+1
